src/dataflow-types/src/client/replicated.rs

```diff
@@ -25,18 +25,77 @@
 use std::collections::{HashMap, HashSet};
 
 use timely::progress::{frontier::MutableAntichain, Antichain};
+use tokio::sync::mpsc::{unbounded_channel, UnboundedReceiver, UnboundedSender};
+use tokio::task::JoinHandle;
 
 use crate::client::Peek;
 use mz_repr::GlobalId;
 
 use super::{ComputeClient, GenericClient};
 use super::{ComputeCommand, ComputeResponse};
 
+// `Client` must have a cancelation-safe `recv` method.
+pub fn spawn_client_tasklet<
+    C: Send + 'static,
+    R: Send + 'static,
+    Client: GenericClient<C, R> + 'static,
+    Name: AsRef<str>,
+    NameClosure: FnOnce() -> Name,
+>(
+    mut client: Client,
+    nc: NameClosure,
+) -> (
+    JoinHandle<()>,
+    UnboundedSender<C>,
+    UnboundedReceiver<Result<R, anyhow::Error>>,
+) {
+    let (cmd_tx, mut cmd_rx) = unbounded_channel();
+    let (response_tx, response_rx) = unbounded_channel();
+    let handle = mz_ore::task::spawn(nc, async move {
+        loop {
+            tokio::select! {
+                m = cmd_rx.recv() => {
+                    match m {
+                        Some(c) => {
+                            // Issues should be detected, and
+                            // reconnect attempted, on the `recv` path.
+                            let _ = client.send(c).await;
+                        },
+                        None => break,
+                    }
+                },
+                m = client.recv() => {
+                    match m {
+                        Ok(None) => break,
+                        Ok(Some(response)) => {
+                            if response_tx.send(Ok(response)).is_err() {
+                                break;
+                            };
+                        }
+                        Err(e) => {
+                            if response_tx.send(Err(e)).is_err() {
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    });
+    (handle, cmd_tx, response_rx)
+}
+
 /// A client backed by multiple replicas.
 #[derive(Debug)]
-pub struct ActiveReplication<C, T> {
-    /// The replicas themselves.
-    replicas: HashMap<String, C>,
+pub struct ActiveReplication<Cmd, Resp, T> {
```

Comment:
`C` and `R` are what we use everywhere else.

```suggestion
pub struct ActiveReplication<C, R, T> {
```